persistent-postgresql/test/JSONTest.hs

```diff
@@ -40,431 +42,654 @@ insert' :: (MonadIO m, PersistStoreWrite backend, BaseBackend backend ~ SqlBacke
         => Value -> ReaderT backend m (Key TestValue)
 insert' = insert . TestValue
 
-(=@=) :: MonadIO m => String -> Bool -> m ()
-s =@= b = liftIO $ assertBool s b
 
 matchKeys :: (Show record, Show (Key record), MonadIO m, Eq (Key record))
-          => String -> [Key record] -> [Entity record] -> m ()
-matchKeys s ys xs = do
-    msg1 =@= (xLen == yLen)
-    forM_ ys $ \y -> msg2 y =@= (y `elem` ks)
+          => [Key record] -> [Entity record] -> m ()
+matchKeys ys xs = do
+    msg1 `assertBoolIO` (xLen == yLen)
+    forM_ ys $ \y -> msg2 y `assertBoolIO` (y `elem` ks)
   where ks = entityKey <$> xs
         xLen = length xs
         yLen = length ys
         msg1 = mconcat
-            [ s, "\nexpected: ", show yLen
+            [ "\nexpected: ", show yLen
             , "\n but got: ", show xLen
-            , "\n[xs: ", show xs
-            , ", ys: ", show ys, "]"
+            , "\n[xs: ", show xs, "]"
+            , "\n[ys: ", show ys, "]"
             ]
         msg2 y = mconcat
-            [ s, ": "
-            , "key \"", show y
+            [ "key \"", show y
             , "\" not in result:\n  ", show ks
             ]
 
+setup :: IO ()
+setup = asIO $ runConn $ do
+    void $ runMigrationSilent jsonTestMigrate
+
+teardown :: IO ()
+teardown = asIO $ runConn $ do
+  cleanDB
+
+shouldBeIO :: (Show a, Eq a, MonadIO m) => a -> a -> m ()
+shouldBeIO x y = liftIO $ shouldBe x y
+
+assertBoolIO :: MonadIO m => String -> Bool -> m ()
+assertBoolIO s b = liftIO $ assertBool s b
+
+edgeCases :: (Monad m, MonadIO m) => ReaderT SqlBackend m (EdgeCaseKeys TestValue)
+edgeCases = do
+  arrList3K <- insert' $ toJSON [toJSON [String "a"], Number 1]
+  arrList4K <- insert' $ toJSON [String "a", String "b", String "c", String "d"]
+  objEmptyK <- insert' $ object ["" .= Number 9001]
+  objFullK  <- insert' $ object ["a" .= Number 1, "b" .= Number 2
+                                , "c" .= Number 3, "d" .= Number 4 ]
+  return $ EdgeCaseKeys {..}
+
+globalKeys :: (Monad m, MonadIO m) => ReaderT SqlBackend m (GlobalKeys TestValue)
+globalKeys = do
+          nullK <- insert' Null
```

Comment:
We should stick to a consistent indentation. The project uses 4 space or 2 space depending on the module (sigh) but nothing uses the 8 space.